File: Algorithm/Greedy-Algorithm/RemoveKDigits.cpp
//leetcode 402

class Solution {
public:
    string removeKdigits(string num, int k) {
        std::vector<int> S;
        std::string result = "";
        for(int i = 0; i < num.length(); i++) {
            int number = num[i] - '0';
            //when stack top smaller than number
            while(S.size() != 0 && S[S.size() - 1] > number && k > 0) {
                S.pop_back();
                k--;
            }
            //avoid the first digit is 0
            if(S.size() != 0 || number != 0) {
                S.push_back(number);
            }
        }
        while(S.size() != 0 && k > 0) {
            S.pop_back();
            k--;
        }
        
        for(int i = 0; i < S.size(); i++) {
            result.append(1, '0' + S[i]);
        }
        if(result == "") {
            result = "0";
        }
        return result;
    }
};

File: Algorithm/GreedyAlgorithm/RemoveKDigits.java
package greedy;

import java.util.Stack;

/**
 * 
 * @description LeetCode: 402, level: Medium
 *
 * https://leetcode.com/problems/remove-k-digits/
 */
public class RemoveKDigits {
    
    public static String solution(String num, int k) {
        StringBuilder result = new StringBuilder();
        
        Stack<Character> stk = new Stack<Character>();
        
        for(int i=0; i < num.length(); i++) {
            char number = num.charAt(i);
            while(!stk.isEmpty() && number < stk.peek() && k > 0) {
                stk.pop();
                k--;
            }
            if(number != '0' || !stk.isEmpty()) {
                stk.push(number);
            }
        }
        
        while(!stk.isEmpty() && k > 0) {
            stk.pop();
            k--;
        }
        
        while(!stk.isEmpty()) {
            result.append(stk.pop());
        }
        
        if(result.length() == 0) {
            return "0";
        }
        
        return result.reverse().toString();
    }

    public static void main(String[] args) {
        String num = "10200";
        System.out.println(solution(num, 1));
    }

}
